File: simlab3/exercise2-1.c
// My code

int retrieveArrayValue(int *array, int len, int index) {
    // Retrieve the element from the array at the specified index
    return array[index];
}

// code produced by CGT

#include <stdio.h>

int getValueFromArray(int *array, int index) {
    // Retrieve the element from the array at the specified index
    return array[index];
}

int main() {
    int array[] = {10, 20, 30, 40, 50};
    int index = 2;
    int value = getValueFromArray(array, index);
    printf("The value at index %d is %d\n", index, value);
    return 0;
}

// my revised code

#include <stdio.h>
#include <stdlib.h>

int retrieveArrayValue(int *array, int len, int index) {
    // Check if array is NULL
    if (array == NULL) {
        fprintf(stderr, "Error: array is NULL\n");
        return -1; // or some other error code/value
    }

    // Check if index is within bounds
    if (index < 0 || index >= len) {
        fprintf(stderr, "Error: index out of bounds\n");
        return -1; // or some other error code/value
    }

    // Retrieve the element from the array at the specified index
    return array[index];
}

int main() {
    int array[] = {10, 20, 30, 40, 50};
    int len = sizeof(array) / sizeof(array[0]);
    int index = 2; // For example, to retrieve the 3rd element
    int value = retrieveArrayValue(array, len, index);
    if (value != -1) { // Check if the value is valid
        printf("The value at index %d is %d\n", index, value);
    }
    return 0;
}

// cgt-generated code revised

#include <stdio.h>

int getValueFromArray(int *array, int len, int index) {
    // Check if index is within bounds
    if (index < 0 || index >= len) {
        fprintf(stderr, "Index out of bounds\n");
        return -1; // Return an error code or handle the error appropriately
    }
    // Retrieve the element from the array at the specified index
    return array[index];
}

int main() {
    int array[] = {10, 20, 30, 40, 50};
    int len = sizeof(array) / sizeof(array[0]);
    int index = 2;
    int value = getValueFromArray(array, len, index);
    if (value != -1) { // Check for error before printing
        printf("The value at index %d is %d\n", index, value);
    }
    return 0;
}

File: simlab3/exercise2-1.py
# My code

def retrieveArrayValue(array, index):
    # Retrieve the element from the array at the specified index
    return array[index]

# code produced by CGT

def get_value_from_array(array, index):
    # Retrieve the element from the array at the specified index
    return array[index]

# Example usage
array = [10, 20, 30, 40, 50]
index = 2
value = get_value_from_array(array, index)
print(f"The value at index {index} is {value}")

# my revised code

def retrieveArrayValue(array, index):
    # Check if array is of valid type
    if not isinstance(array, (list, tuple)):
        raise TypeError("array must be a list or tuple")
    
    # Check if index is of valid type
    if not isinstance(index, int):
        raise TypeError("index must be an integer")

    # Check if index is within bounds
    if index < 0 or index >= len(array):
        raise IndexError("index out of bounds")

    # Retrieve the element from the array at the specified index
    return array[index]

# Example usage
array = [10, 20, 30, 40, 50]
index = 2
try:
    value = retrieveArrayValue(array, index)
    print(f"The value at index {index} is {value}")
except (TypeError, IndexError) as e:
    print(f"Error: {e}")

# cgt-generated revised code

def get_value_from_array(array, index):
    # Check if index is within bounds
    if index < 0 or index >= len(array):
        raise IndexError("Index out of bounds")
    # Retrieve the element from the array at the specified index
    return array[index]

# Example usage
array = [10, 20, 30, 40, 50]
index = 2
try:
    value = get_value_from_array(array, index)
    print(f"The value at index {index} is {value}")
except IndexError as e:
    print(e)

